File: sdk/ruby/lib/bsdkrun/binary.rb
# frozen_string_literal: true

require "pathname"

module Bsdkrun
  # Locates the +bsdkrun+ binary on the host and caches the result.
  #
  # Resolution order (first match wins):
  #   1. an explicit override set via {Bsdkrun.binary_path=}
  #   2. the +BSDKRUN_BIN+ environment variable
  #   3. +bsdkrun+ on +PATH+
  #   4. an in-repo dev build: +<repo>/target/release/bsdkrun+ then +debug+
  module Binary
    @override = nil
    @resolved = nil

    class << self
      # Force the SDK to use a specific +bsdkrun+ binary, bypassing discovery.
      #
      # @param path [String, nil]
      # @return [void]
      def override=(path)
        @override = path
        @resolved = nil
      end

      # @return [String, nil] the current explicit override, if any.
      attr_reader :override

      # Reset cached discovery state (mainly for tests).
      # @return [void]
      def reset!
        @resolved = nil
        @override = nil
      end

      # Resolve (and cache) the path to the +bsdkrun+ binary.
      #
      # @return [String] the resolved absolute path or bare command name.
      # @raise [BinaryNotFound] when nothing matched.
      def resolve
        return @resolved if @resolved

        searched = candidates
        searched.each do |candidate|
          if path_like?(candidate)
            return @resolved = candidate if File.exist?(candidate)
          else
            found = which(candidate)
            return @resolved = found if found
          end
        end
        raise BinaryNotFound, searched
      end

      private

      def path_like?(candidate)
        candidate.include?(File::SEPARATOR) ||
          (File::ALT_SEPARATOR && candidate.include?(File::ALT_SEPARATOR))
      end

      # Candidate locations, in priority order.
      def candidates
        out = []
        out << @override if @override

        env = ENV["BSDKRUN_BIN"]
        out << env if env && !env.empty?

        # A `bsdkrun` already on PATH wins over in-repo builds.
        on_path = which("bsdkrun")
        out << on_path if on_path

        # This file lives at sdk/ruby/lib/bsdkrun/binary.rb, so the repo root is
        # four levels up. Prefer a release build, fall back to debug.
        repo_root = File.expand_path("../../../..", __dir__)
        out << File.join(repo_root, "target", "release", "bsdkrun")
        out << File.join(repo_root, "target", "debug", "bsdkrun")
        out
      end

      # Cross-platform PATH lookup for an executable.
      def which(name)
        exts =
          if File::ALT_SEPARATOR
            (ENV["PATHEXT"] || ".EXE;.CMD;.BAT").split(";")
          else
            [""]
          end
        (ENV["PATH"] || "").split(File::PATH_SEPARATOR).each do |dir|
          next if dir.empty?

          exts.each do |ext|
            full = File.join(dir, name + ext)
            return full if File.file?(full) && File.executable?(full)
          end
        end
        nil
      end
    end
  end
end

File: sdk/clojure/src/bsdkrun/binary.clj
(ns bsdkrun.binary
  "Locates the `bsdkrun` binary on the host and caches the result.

  Resolution order (first match wins):

    1. an explicit override set via [[set-override!]]
    2. the `BSDKRUN_BIN` environment variable
    3. `bsdkrun` on `PATH`
    4. an in-repo dev build: `<repo>/target/release/bsdkrun`, then
       `<repo>/target/debug/bsdkrun`

  [[candidates]] and [[resolve]] both take an optional opts map
  (`:override`, `:bsdkrun-bin`, `:path`, `:repo-root`) so the discovery logic
  is unit-testable without mutating real process/environment state — the JVM
  has no supported, portable way to change `System/getenv` for the current
  process, unlike Ruby's `ENV[...]=`. Any key left out of the map falls back
  to the real host state (the override atom, `$BSDKRUN_BIN`, `$PATH`, and the
  monorepo root inferred from this namespace's own source location)."
  (:refer-clojure :exclude [reset! resolve])
  (:require [clojure.java.io :as io]
            [clojure.string :as str]
            [bsdkrun.errors :as errors])
  (:import [java.io File]))

(def ^:private state (atom {:override nil :resolved nil}))

(defn set-override!
  "Force the SDK to use a specific `bsdkrun` binary, bypassing discovery."
  [path]
  (swap! state assoc :override path :resolved nil)
  nil)

(defn override
  "The current explicit override, if any."
  []
  (:override @state))

(defn reset!
  "Reset cached discovery state and any override (mainly for tests)."
  []
  (swap! state (constantly {:override nil :resolved nil}))
  nil)

(defn- path-like?
  [^String candidate]
  (str/includes? candidate File/separator))

(defn- executable-file?
  [^File f]
  (and (.isFile f) (.canExecute f)))

(defn- which
  "Cross-platform PATH lookup for an executable name. `path-env` is a raw
  `PATH`-style string (entries separated by `File/pathSeparator`); defaults
  to the real `$PATH`. Returns the resolved absolute path, or nil."
  ([name] (which name (System/getenv "PATH")))
  ([name path-env]
   (let [sep (java.util.regex.Pattern/quote File/pathSeparator)
         dirs (str/split (or path-env "") (re-pattern sep))]
     (some (fn [dir]
             (when (seq dir)
               (let [f (io/file dir name)]
                 (when (executable-file? f) (.getPath f)))))
           dirs))))

(defn- resource-dir
  "The directory containing this namespace's own source file on disk, or nil
  if it can't be resolved to a plain file (e.g. running from inside a jar)."
  []
  (when-let [res (io/resource "bsdkrun/binary.clj")]
    (when (= "file" (.getProtocol res))
      (.getParentFile (io/file (.toURI res))))))

(defn- repo-root-dir
  "The monorepo root, four directories up from this namespace's source
  directory: `sdk/clojure/src/bsdkrun` -> `sdk/clojure/src` -> `sdk/clojure`
  -> `sdk` -> repo root. Same depth Ruby's `binary.rb` walks up from
  `sdk/ruby/lib/bsdkrun`. Returns nil when [[resource-dir]] can't be
  determined."
  []
  (when-let [dir (resource-dir)]
    (-> ^File dir .getParentFile .getParentFile .getParentFile .getParentFile)))

(defn- default-opts
  []
  {:override (:override @state)
   :bsdkrun-bin (let [v (System/getenv "BSDKRUN_BIN")] (when (seq v) v))
   :path (System/getenv "PATH")
   :repo-root (repo-root-dir)})

(defn candidates
  "Candidate `bsdkrun` binary locations, in priority order, as a vector of
  path strings. A pure function of `opts` (see the namespace docstring for
  the supported keys and their real-host defaults)."
  ([] (candidates {}))
  ([opts]
   (let [{:keys [override bsdkrun-bin path repo-root]} (merge (default-opts) opts)
         on-path (which "bsdkrun" path)]
     (cond-> []
       override (conj override)
       bsdkrun-bin (conj bsdkrun-bin)
       on-path (conj on-path)
       repo-root (conj (.getPath (io/file repo-root "target" "release" "bsdkrun")))
       repo-root (conj (.getPath (io/file repo-root "target" "debug" "bsdkrun")))))))

(defn resolve
  "Resolve (and cache) the path to the `bsdkrun` binary.

  Throws `ex-info` (kind `:bsdkrun/binary-not-found`, see `bsdkrun.errors`)
  listing everything that was searched if nothing matched."
  ([] (resolve {}))
  ([opts]
   (or (:resolved @state)
       (let [merged (merge (default-opts) opts)
             searched (candidates merged)
             found (some (fn [candidate]
                            (if (path-like? candidate)
                              (when (.exists (io/file candidate)) candidate)
                              (which candidate (:path merged))))
                          searched)]
         (if found
           (do (swap! state assoc :resolved found) found)
           (throw (errors/binary-not-found searched)))))))
